File: src/main/java/it/unibo/crabinv/Controller/core/GameLoopController.java
package it.unibo.crabinv.Controller.core;

import it.unibo.crabinv.Model.core.GameSnapshot;

/**
 * Controls the passing of time and creates the snapshot to be passed onto the renderer
 */
public interface GameLoopController {
    /**
     * @return the duration in milliseconds of a tick
     */
    long getTickDurationMillis();

    /**
     * @return the currently total accumulated milliseconds,
     * needed to calculate the next step
     */
    long getAccumulatedMillis();

    /**
     * @return the total of ticks since the start of the engine
     */
    long getTotalElapsedTicks();

    /**
     * Calculates the game engine next step snapshot to be passed onto the renderer.
     * Advances the logic of the game engine of the calculated step.
     * @param frameElapsedMillis the time accumulated between two frames
     * @return the calculated snapshot
     */
    GameSnapshot step(long frameElapsedMillis);

    /**
     * @return the latest snapshot to be rendered, without any step advancing
     */
    GameSnapshot getLatestSnapshot();

}

File: src/main/java/it/unibo/crabinv/controller/core/gameloop/GameLoopController.kt
package it.unibo.crabinv.controller.core.gameloop

import it.unibo.crabinv.model.core.snapshot.GameSnapshot

/**
 * Controls the passing of time and creates the [GameSnapshot] to be rendered.
 */
interface GameLoopController {
    /**
     * @return the duration in milliseconds of a tick.
     */
    val tickDurationMillis: Long

    /**
     * @return the currently total accumulated milliseconds, needed to calculate the next step.
     */
    val accumulatedMillis: Long

    /**
     * @return the total of ticks since the start of the engine.
     */
    val totalElapsedTicks: Long

    /**
     * Calculates the game engine next step [GameSnapshot] to be passed onto the renderer.
     * Advances the game logic by the calculated step.
     * @param frameElapsedMillis the milliseconds accumulated between two frames, must be positive.
     * @return the calculated [GameSnapshot]
     */
    fun step(frameElapsedMillis: Long): GameSnapshot?

    /**
     * @return the latest [GameSnapshot]
     */
    val latestSnapshot: GameSnapshot?

    /**
     * Requests the [it.unibo.crabinv.model.core.engine.GameEngine] to pause the game.
     */
    fun pause()

    /**
     * Requests the [it.unibo.crabinv.model.core.engine.GameEngine] to resume the game.
     */
    fun resume()
}
